zerver/views/auth.py

```diff
@@ -460,16 +460,13 @@ def start_social_login(request: HttpRequest, backend: str, extra_arg: Optional[s
                          .format(extra_arg))
             return redirect_to_config_error("saml")
         extra_url_params = {'idp': extra_arg}
-    if (backend == "github") and not (settings.SOCIAL_AUTH_GITHUB_KEY and
-                                      settings.SOCIAL_AUTH_GITHUB_SECRET):
-        return redirect_to_config_error("github")
-    if (backend == "google") and not (settings.SOCIAL_AUTH_GOOGLE_KEY and
-                                      settings.SOCIAL_AUTH_GOOGLE_SECRET):
-        return redirect_to_config_error("google")
-    if (backend == "gitlab") and not (settings.SOCIAL_AUTH_GITLAB_KEY and
-                                      settings.SOCIAL_AUTH_GITLAB_SECRET):
-        return redirect_to_config_error("gitlab")
-    # TODO: Add a similar block for AzureAD.
+    backends = ["github", "google", "gitlab"]
+    # TODO: Add AzureAD also.
+    for backend in backends:
+        key_setting = "SOCIAL_AUTH_" + backend.upper() + "_KEY"
+        secret_setting = "SOCIAL_AUTH_" + backend.upper() + "_SECRET"
+        if not (getattr(settings, key_setting) and getattr(settings, secret_setting)):
+            return redirect_to_config_error(backend)
```

Comment:
This can be improved further; you should be able to just do a loop through `EXTERNAL_AUTHENTICATION_METHODS`.  SAML is already handled above, and if there's any that we don't want this behavior to happen for (maybe `ZulipRemoteUserBackend`), we should just have a `custom_config_error=True` flag on the auth backend class to configure that behavior.

Also, doing that loop will allow us to just use the CLIENT_KEY_SETTING type setting values (which I suppose we might need to move from the test base classes to the auth backend classes, but that seems like a good idea anyway).

I'm still happy to merge this, but it'd be great if you added another commit doing this.

Reply:
Also, following the 3rd commit, I wonder if we can just do the `render_markdown_path` in the caller, rather than in the template, and avoid the ugly line-wrapping business.

Reply:
>This can be improved further; you should be able to just do a loop through EXTERNAL_AUTHENTICATION_METHODS
so, do we need something like `check_config` funcion that saml has? I think that would be good and we do not need to use CLIENT_KEY_SETTING in a loop if we make `check_config` for every backend auth.

Reply:
> Also, following the 3rd commit, I wonder if we can just do the `render_markdown_path` in the caller, rather than in the template, and avoid the ugly line-wrapping business.

I'm not sure if we can do that. @mateuszmandera You might be able to answer this.
Even if we can I think we have to make a different `.md` file for every backend auth having some duplicate text.